source/s3_util.c

```diff
@@ -488,6 +488,74 @@ int aws_s3_parse_content_length_response_header(
     return result;
 }
 
+int aws_s3_parse_request_range_header(
+    struct aws_http_headers *request_headers,
+    bool *out_initial_message_has_start_range,
+    bool *out_initial_message_has_end_range,
+    uint64_t *out_initial_start_range,
+    uint64_t *out_initial_end_range) {
+
+    AWS_PRECONDITION(request_headers);
+    AWS_PRECONDITION(out_initial_message_has_start_range);
+    AWS_PRECONDITION(out_initial_message_has_end_range);
+    AWS_PRECONDITION(out_initial_start_range);
+    AWS_PRECONDITION(out_initial_end_range);
+
+    struct aws_byte_cursor range_header_value;
+
+    if (aws_http_headers_get(request_headers, g_range_header_name, &range_header_value)) {
+        // TODO: Just assert?
+        return aws_raise_error(AWS_ERROR_S3_INVALID_RANGE_HEADER);
+    }
+    struct aws_byte_cursor range_header_start = aws_byte_cursor_from_c_str("bytes=");
+
+    int result = AWS_OP_ERR;
+    /* verify bytes= */
+    if (aws_byte_cursor_starts_with(&range_header_value, &range_header_start)) {
+        aws_byte_cursor_advance(&range_header_value, range_header_start.len);
+        struct aws_byte_cursor substr = {0};
+        /* parse start range */
+        if (!aws_byte_cursor_next_split(&range_header_value, '-', &substr)) {
+            goto done;
+        }
+        if (substr.len > 0) {
+            if (aws_byte_cursor_utf8_parse_u64(substr, out_initial_start_range)) {
+                goto done;
+            }
+            *out_initial_message_has_start_range = true;
+        }
+        /* parse end range */
+        if (!aws_byte_cursor_next_split(&range_header_value, '-', &substr)) {
+            goto done;
+        }
+        if (substr.len > 0) {
+            if (aws_byte_cursor_utf8_parse_u64(substr, out_initial_end_range)) {
+                goto done;
+            }
+            *out_initial_message_has_end_range = true;
+        }
+
+        /* verify there is not anything extra */
+        if (aws_byte_cursor_next_split(&range_header_value, '-', &substr)) {
+            goto done;
+        }
+
+        /* verify that start-range <= end-range */
+        if (out_initial_message_has_end_range && out_initial_start_range > out_initial_end_range) {
```

Comment:
When we perform a `HeadRequest` with `startRange > endRange` such as `bytes=20-10`, S3 does not add a `Content-Range` header in the `HeadRequest` response, and `Content-Length` reflects the total length of the object. The CLI downloads the entire object. The current CRT behavior is that it fails with `AWS_ERROR_S3_MISSING_CONTENT_RANGE_HEADER` at https://github.com/awslabs/aws-c-s3/blob/main/source/s3_auto_ranged_get.c#L510C24-L510C66. According to the HTTP [spec](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.1-6), “An int-range is invalid if the last-pos value is present and less than the first-pos.” . This changes it to `AWS_ERROR_S3_INVALID_RANGE_HEADER` error at request time.